app/src/main/java/com/codingblocks/cbonlineapp/viewmodels/VideoPlayerViewModel.kt

```diff
@@ -0,0 +1,74 @@
+package com.codingblocks.cbonlineapp.viewmodels
+
+import androidx.lifecycle.ViewModel
+import com.codingblocks.cbonlineapp.database.CourseDao
+import com.codingblocks.cbonlineapp.database.CourseRunDao
+import com.codingblocks.cbonlineapp.database.DoubtsDao
+import com.codingblocks.cbonlineapp.database.NotesDao
+import com.codingblocks.cbonlineapp.extensions.retrofitCallback
+import com.codingblocks.onlineapi.Clients
+import com.codingblocks.onlineapi.models.ContentsId
+import com.codingblocks.onlineapi.models.DoubtsJsonApi
+import com.codingblocks.onlineapi.models.Notes
+import com.codingblocks.onlineapi.models.RunAttemptsId
+import com.google.gson.JsonObject
+import retrofit2.Response
+
+class VideoPlayerViewModel(
+    var doubtsDao: DoubtsDao,
+    var notesDao: NotesDao,
+    var courseDao: CourseDao,
+    var runDao: CourseRunDao
+) : ViewModel() {
+    lateinit var attemptId: String
+    lateinit var contentId: String
+    lateinit var sectionId: String
+    lateinit var videoId: String
+    var download: Boolean = false
+
+    var playWhenReady = false
+    var currentOrientation: Int = 0
+
+    fun getOtp(onVideoCompleteListener: OnVideoCompleteListener) {
+        Clients.api.getOtp(videoId, sectionId, attemptId)
+            .enqueue(retrofitCallback { _, response ->
+                response?.let {
+                    if (it.isSuccessful) {
+                        onVideoCompleteListener.onComplete(it)
+                    } else
+                        onVideoCompleteListener.onFailed()
+                }
+            })
+    }
+
+    @Suppress("UNCHECKED_CAST")
+    fun createDoubt(doubt: DoubtsJsonApi, onVideoCompleteListener: OnVideoCompleteListener) {
+        doubt.postrunAttempt = RunAttemptsId(attemptId)
+        doubt.contents = ContentsId(contentId)
+        Clients.onlineV2JsonApi.createDoubt(doubt)
+            .enqueue(retrofitCallback { _, response ->
+                if (response?.isSuccessful == true)
+                    onVideoCompleteListener.onComplete(response as Response<JsonObject>)
+                else
+                    onVideoCompleteListener.onFailed()
+            })
+    }
+
+    @Suppress("UNCHECKED_CAST")
+    fun createNote(note: Notes, onVideoCompleteListener: OnVideoCompleteListener) {
+        note.content = ContentsId(contentId)
+        note.runAttempt = RunAttemptsId(attemptId)
+        Clients.onlineV2JsonApi.createNote(note)
+            .enqueue(retrofitCallback { _, response ->
+                if ((response as Response<Notes>).isSuccessful)
+                    onVideoCompleteListener.onComplete(response as Response<JsonObject>)
+                else
+                    onVideoCompleteListener.onFailed()
+            })
+    }
+}
+
+interface OnVideoCompleteListener {
+    fun onComplete(response: Response<JsonObject>)
```

Comment:
And why JsonObject obviously it won't work

Reply:
removed

Reply:
im updating the course activity, give your feedback on the changes. If it gets accepted I shall go on and perform the same on the Video Activity

Reply:
First of all they are Gson Response not Json Response and secondly it shouldn't be like that it should be at all you should save the dao here itself and if error show an error message using a boolean livedata variable and snackbar message

Reply:
I didn't change the Video Activity as of now. I've updated only the Course Activity and it's ViewModel with the livedata and observers with the mutable boolean variables. I will make changes to the VideoClass once you confirm the refactor for One Class

Reply:
Btw I don't get what you mean by Gson Response. It is a Retrofit Response of Gson JsonObject right?

Reply:
No it's GSON see client class in onlineapi. Module

Reply:
@aggarwalpulkit596 Check this Sir.

The Call
<img width="348" alt="Screenshot 2019-05-30 at 2 37 02 PM" src="https://user-images.githubusercontent.com/11988517/58622140-8fe9d280-82e8-11e9-8f03-45d79dd0a7ff.png">

The GSON JsonObject import
<img width="360" alt="Screenshot 2019-05-30 at 2 37 10 PM" src="https://user-images.githubusercontent.com/11988517/58622137-8d877880-82e8-11e9-9b5d-3c0fc6de36ec.png">

Reply:
Not every call is a JsonObject notes are of type GSON

Reply:
There are two different interfaces onlineRESTAPI and onlineJSONAPI so do your work accordingly